bryce-market-maker.md

```diff
@@ -0,0 +1,122 @@
+# Design a Market Maker
+
+A Market Maker is a company that provides liquidity to the market, by
+buying an asset when there is an offer to sell an asset for a certain
+amount below the asset's real worth, and selling an asset when there is
+an offer to buy an asset for a certain amount above the asset's real
+worth.
+
+For example, let's say AAPL is selling for $100, and we will buy any
+AAPL sold for $99, and will sell AAPL to any buyer willing to buy AAPL
+for more than $101.
+
+Design an algorithm for a market maker, and design a system to scale it
+properly.
+
+## Algorithm
+
+Write some code to create an algorithm that would maximize the amount of
+money, assuming that nobody else buys and sells.
+
+What happens if this becomes an n-sided market, where there are many
+market makers? How does this change the problem?
+
+Assume you have a limited amount of capital, which you don't want to
+expend all at once (since there's volatility in asset prices, which may
+go up +-10% a day), and there might be a larger amount of sell offers
+than buy offers, for a particular asset. What happens if a stock goes up
+too much, or down too much in a given day?
+
+
+### Algorithm thoughts
+We need a greedy algorithm that has fast access going through the existing data. We also need to concern ourselves with how are we getting access to this market data? 
+We will be looking at the market one state at a time. I would say just find the maximum buy order at a given time and go for it, but keep in mind other people will be 
+competing with us to sell to customers. So I reccomend we find the first existing buy order that gives us that 1% profit and execute it before our competition beats us. 
+This will accomplish 2 things.
+1.   we get to the resource faster than someone with a more greedy profit margin
+2. we will have very fast turnaround time and allow us to execute more orders in each batch, thus making profit via quantity of sales, rather than one big payday. 
+
+## System Design
+Design a system that listens to a server that sends requests, records
+them to a database, and executes or passes on each offer, based on if it
+is being sold for a 1% discount or if someone is willing to buy at a 1%
+premium.
+
+# Step 1: Outline usecases, constraints, and assumptions
+- Q1. Is this system more read or more write heavy?
+- A1. probably pretty balanced in the ratio because we will show many stock prices to many users as well as store a history of the orders
+- Q2. What is the best data representation of stock information? the scc will want to have access to all of our trades so i reccomend we split the data into two layers
+1. The customer layer which stores a minimal amount of information to allow for speed
+2. a more reliable timeseries based layer we can use to turn in tax information and have our quants do analysis of our profits and maybe tweak our algortihm accordingly
```

Comment:
What advantages would a timeseries based layer have over a regular rdbms?

Reply:
we probably could just use some sort of appendonly structure instead now that i think more deeply about it. Because we will never modify this part of the product